### PR TITLE
searchmatch

### PR DESCRIPTION
**Pull request by vasukalariya**
# Description
Summarize the change.
SearchMatch SPL command

Fixes #<issue-number> (link all the GitHub issues this addresses)

# Testing
Describe how you tested this code. How can the reviewers reproduce your tests?
```

app_name = "Bracecould" | eval n = if(searchmatch("city=Orlando"), "yes", "no")

app_name = "Bracecould" | eval n = if(searchmatch("Orlando"), "yes", "no")

app_name = "Bracecould" | eval n = if(searchmatch("city=O*"), "yes", "no")

app_name = "Bracecould" | eval n = if(searchmatch("first_name=A*e"), "yes", "no")

app_name = "Bracecould" | eval n = if(searchmatch("group=group?0"), "yes", "no")

* | eval n = if(searchmatch("gender=????"), "yes", "no")

city=Boston | stats count AS Count BY http_status | eval n = if(searchmatch("http_status=200"), "yes", "no")

city=Boston | stats count AS Count BY http_status | eval n = if(searchmatch("http_status=?00"), "yes", "no")

city=Boston | stats count AS Count BY http_status | eval n = if(searchmatch("Count=*"), "yes", "no")

* | eval x="hi" | eval y="goodbye" | eval test=if(searchmatch("x=hi y=*"), "yes", "no")

```

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [x] I have self-reviewed this PR.
- [x] I have removed all print-debugging and commented-out code that should not be merged.
- [x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/ast/spl/spl.peg [hunk 1]
```diff
@@ -295,6 +295,39 @@ func getRexColNames(pattern string) ([]string, error) {
...
+	for _, kvPair := range kvPairs {
+		parts := strings.Split(kvPair, "=")
+		if len(parts) == 1 {
+            if len(kvPairs) == 1 {
+                fieldList = append(fieldList, "*")
+			    return fieldList, nil
+            } else {
+                return []string{}, fmt.Errorf("Spl peg: Invalid syntax for searchmatch") 
+            }
+		}
```
[the remark this answers is not in the document]
Comment (AndrewHess) [reply pkg/ast/spl/spl.peg:315]: Also where did you find documentation for how this part is supposed to work? I only see https://docs.splunk.com/Documentation/Splunk/9.1.1/SearchReference/ConditionalFunctions#searchmatch.28.26lt.3Bsearch_str.26gt.3B.29 which doesn't give much info

File: pkg/ast/spl/spl.peg [hunk 2]
```diff
@@ -2065,6 +2098,25 @@ EvalComparisonExpr <- (op:("isbool" / "isint" / "isstr" / "isnull" / "isnotnull"
...
+    }
+
+    fieldsList, err := SearchMatchGetFields(valExpr.StringExpr.RawString)
+    if err != nil {
+        return nil, err
+    }
+
+    valExpr.StringExpr.StringExprMode = structs.SEMFieldList
+    valExpr.StringExpr.FieldList = fieldsList
+    expr.LeftValue = valExpr
```
Comment (AndrewHess) [inline pkg/ast/spl/spl.peg:2118]: nit: instead of
```
    if opNameStr != "searchmatch" {
        return expr, nil
    }
```
And then handling the logic for `searchmatch`, I think it would be easier to read (especially if we have to add more special cases later) if you do
```
if opNameStr == "searchmatch" {
    // handle the logic here
}
```
Or you could move the searchmatch logic to a different function if you want to

File: pkg/ast/spl/tests/splParser_test.go [hunk 3]
```diff
@@ -8240,3 +8242,116 @@ func performEventCountTest(t *testing.T, query string, expectedIndices []string,
...
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
```
Comment (AndrewHess) [inline pkg/ast/spl/tests/splParser_test.go:8256]: Assuming there isn't a function to do this in a library we're using already, this function probably belongs in sliceutils.go. If you want, you could even make this a generic function so that we could compare any two slices containing `Comparable` types so that we could reuse it e.g., to compare slices of ints; but you don't have to write the generic function

File: pkg/segment/structs/evaluationstructs.go [hunk 4]
```diff
@@ -359,17 +362,19 @@ func (self *SortExpr) ReleaseProcessedSegmentsLock() {
 
 func findNullFields(fields []string, fieldToValue map[string]utils.CValueEnclosure) ([]string, error) {
 	nullFields := []string{}
+	var err error
 	for _, field := range fields {
 		val, exists := fieldToValue[field]
 		if !exists {
-			return []string{}, fmt.Errorf("findNullFields: Expression has a field for which value is not present")
+			err = fmt.Errorf("findNullFields: Expression has a field %v for which value is not present", field)
+			continue
```
Comment (AndrewHess) [inline pkg/segment/structs/evaluationstructs.go:370]: What's the purpose of this change? Why don't we want to return on the first error?

File: pkg/segment/structs/evaluationstructs.go [hunk 5]
```diff
@@ -408,6 +413,66 @@ func (self *RexExpr) GetNullFields(fieldToValue map[string]utils.CValueEnclosure
 	return findNullFields(self.GetFields(), fieldToValue)
 }
 
+func checkStringInFields(searchStr string, fieldToValue map[string]utils.CValueEnclosure) (bool, error) {
+	for _, v := range fieldToValue {
+		val, err := v.GetString()
```
Comment (AndrewHess) [inline pkg/segment/structs/evaluationstructs.go:418]: nit: rename `v`. It looks like it's a `CValueEnclosure`, so a name indicating that would be good

Also maybe something like `stringValue` would be a little easier to understand than `val`. What do you think?

File: pkg/ast/spl/spl.peg (hunk 1 above)
[the remark this answers is not in the document]
Comment (vasukalariya) [reply pkg/ast/spl/spl.peg:315]: I got the hint about this from https://community.splunk.com/t5/Splunk-Search/Can-t-we-use-NOT-in-searchmatch-function/m-p/465526#M131154
I am not too sure but it seems like it that it searches from all the fields if single string is present.

File: pkg/ast/spl/spl.peg [hunk 6]
```diff
@@ -295,6 +295,39 @@ func getRexColNames(pattern string) ([]string, error) {
     return rexColNames, nil
 }
 
+func SearchMatchGetFields(searchStr string) ([]string, error) {
+
+    kvPairs := strings.Fields(searchStr)
```
[the remark this answers is not in the document]
Comment (vasukalariya) [reply pkg/ast/spl/spl.peg:300]: Documentation is quite poor for searchmatch, mostly I have not seen any examples with space or quotes in splunk community. So I am using the getFields. Also even if they want to enter a quote or space, there is no specific example or occurence of how they will escape it.

File: pkg/segment/structs/evaluationstructs.go (hunk 4 above)
Comment (vasukalariya) [reply pkg/segment/structs/evaluationstructs.go:370]: I want to know the null fields and evaluate those to false. There might be some cases where we might want to know which are the null fields irrespective or if a field that was supposed to have a value does not have it. This function was early exiting so I think this would be a better approach.

File: cicd/ingest.csv [hunk 7]
```diff
@@ -130,4 +130,12 @@ city=Boston | stats count AS Count BY http_method | eval newField=lower(http_met
...
\ No newline at end of file
+"city=Boston | stats count BY http_status | eval original_date_string=""17:19:01 UTC"" | eval original_parsed_time=strptime(original_date_string, ""%H:%M:%S %Z"") | eval formatted_time=strftime(original_parsed_time*1000, ""%a %m/%d/%Y %H:%M:%S %Z"") | eval reparsed_time=strptime(formatted_time, ""%a %m/%d/%Y %H:%M:%S %Z"") | eval res=if(original_parsed_time=reparsed_time ,""pass"" ,""fail"")",now-1d,now,*,group:res:400,eq,"pass",Splunk QL
+"city=Boston | stats count AS Count BY http_status | eval match = if(searchmatch(""http_status=200""), ""yes"", ""no"")",now-1d,now,*,group:match:200,eq,"yes",Splunk QL
+"city=Boston | stats count AS Count BY http_status | eval match = if(searchmatch(""http_status=?00""), ""yes"", ""no"")",now-1d,now,*,group:match:500,eq,"yes",Splunk QL
+"city=Boston | stats count AS Count BY http_status | eval match = if(searchmatch(""http_status=?00""), ""yes"", ""no"")",now-1d,now,*,group:match:400,eq,"yes",Splunk QL
+"city=Boston | stats count AS Count BY http_status | eval match = if(searchmatch(""http_status=?00""), ""yes"", ""no"")",now-1d,now,*,group:match:200,eq,"yes",Splunk QL
```
[the remark this answers is not in the document]
Comment (vasukalariya) [reply cicd/ingest.csv]: addressed in latest commit

File: pkg/ast/spl/spl.peg (hunk 1 above)
[the remark this answers is not in the document]
Comment (vasukalariya) [reply pkg/ast/spl/spl.peg]: addressed in latest commit

File: pkg/ast/spl/spl.peg (hunk 6 above)
[the remark this answers is not in the document]
Comment (vasukalariya) [reply pkg/ast/spl/spl.peg:300]: addressed in latest commit

File: pkg/ast/spl/spl.peg (hunk 2 above)
Comment (vasukalariya) [reply pkg/ast/spl/spl.peg]: addressed in latest commit

File: pkg/ast/spl/tests/splParser_test.go (hunk 3 above)
Comment (vasukalariya) [reply pkg/ast/spl/tests/splParser_test.go]: addressed in latest commit

File: pkg/segment/structs/evaluationstructs.go (hunk 4 above)
Comment (vasukalariya) [reply pkg/segment/structs/evaluationstructs.go]: addressed in latest commit

File: pkg/segment/structs/evaluationstructs.go (hunk 5 above)
Comment (vasukalariya) [reply pkg/segment/structs/evaluationstructs.go]: addressed in latest commit

**Review by AndrewHess — APPROVED** [review]
